src/idiomatic/leveraging-the-type-system/extension-traits/method-resolution-conflicts.md

```diff
@@ -0,0 +1,79 @@
+---
+minutes: 15
+---
+
+# Method Resolution Conflicts
+
+What happens when you have a name conflict between an inherent method and an
+extension method?
+
+```rust
+mod ext {
+    pub trait StrExt {
+        fn trim_ascii(&self) -> &str;
+    }
+
+    impl StrExt for &str {
+        fn trim_ascii(&self) -> &str {
+            self.trim_start_matches(|c: char| c.is_ascii_whitespace())
+        }
+    }
+}
+
+pub use ext::StrExt;
+// Which `trim_ascii` method is invoked?
+// The one from `StrExt`? Or the inherent one from `str`?
+assert_eq!(" dad ".trim_ascii(), "dad");
+```
+
+<details>
+
+- The foreign type may, in a newer version, add a new inherent method with the
+  same name of our extension method.
+
+  Survey the class: what do the students think will happen in the example above?
+  Will there be a compiler error? Will one of the two methods be given higher
+  priority? Which one?
+
+  Add a `panic!("Extension trait")` in the body of `StrExt::trim_ascii` to
+  clarify which method is being invoked.
+
+- [Inherent methods have higher priority than trait methods][1], _if_ they have
+  the same name and the **same receiver**, e.g. they both expect `&self` as
+  input. The situation becomes more nuanced if the use a **different receiver**,
+  e.g. `&mut self` vs `&self`.
+
+  Change the signature of `StrExt::trim_ascii` to
+  `fn trim_ascii(&mut self) -> &str` and modify the invocation accordingly:
+
+  ```rust
+  assert_eq!((&mut " dad ").trim_ascii(), "dad");
+  ```
+
+  Now `StrExt::trim_ascii` is invoked, rather than the inherent method, since
+  `&mut self` is a more specific receiver than `&self`, the one used by the
```

Comment:
Is "more specific" defined somewhere?

Reply:
No, it's better to keep using the priority terminology. Reworded in [1670b64](https://github.com/google/comprehensive-rust/pull/2812/commits/1670b6468c86be8bfb96b53452ed14be5457f598)